### PR TITLE
feat(rust): add `comments` option

### PR DESCRIPTION
**Pull request by hyf0**
### Description

Since we have a feature-complete js compiler, instead of using `legalComments`, we gonna have a more general name `comments` for controlling behavior related to comments. @yyx990803 cc

This PR adds two option:
- Preserve all comments as possible
- Remove all comments

I'll add legal-comments-related feature in later PRs.

### CODE REVIEW

**Comment by hyf0** [comment]
* **#2624** <a href="https://app.graphite.dev/github/pr/rolldown/rolldown/2624?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2615** <a href="https://app.graphite.dev/github/pr/rolldown/rolldown/2615?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2614** <a href="https://app.graphite.dev/github/pr/rolldown/rolldown/2614?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @hyf0 and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: crates/rolldown/tests/rolldown/function/comments/preserve/artifacts.snap [hunk 1]
```diff
@@ -0,0 +1,16 @@
+---
+source: crates/rolldown_testing/src/integration_test.rs
+snapshot_kind: text
+---
+# Assets
+
+## main.js
+
+```js
```
Comment (hyf0) [inline crates/rolldown/tests/rolldown/function/comments/preserve/artifacts.snap:9]: I'm not sure why it doesn't work. I check the passing value is indeed `comments: true`, so let's not block merging this PR due to it doesn't work.

@Boshen cc

Comment (hyf0) [reply crates/rolldown/tests/rolldown/function/comments/preserve/artifacts.snap:9]: Does `print_all_comments` should be superset of `print_legal_comments_only` ?

It's wired that similar case is passed for `print_legal_comments_only`.

https://github.com/rolldown/rolldown/pull/2615/files#diff-dc8ff37288ec95c0339a59b4e355c861f7ad41ad1f24af5025a5548d369fbc93

**Comment by IWANABETHATGUY** [comment]
I tested it, here is a example and it works: 
```
/* #__PURE__ */ foo(a, b);
```

**Comment by IWANABETHATGUY** [comment]
> I tested it, here is a example and it works:
> 
> ```
> /* #__PURE__ */ foo(a, b);
> ```

The `comments` option in codegen works well but may not work as your expectation, comments is not guaranteed to be generated in any position, both esbuild and swc can't do that.

**Comment by hyf0** [comment]
I'll tweak the tests later, let's merge this PR since there aren't blockers.

**Comment by hyf0** [comment]
> > I tested it, here is a example and it works:
> > ```
> > /* #__PURE__ */ foo(a, b);
> > ```
> 
> The `comments` option in codegen works well but may not work as your expectation, comments is not guaranteed to be generated in any position, both esbuild and swc can't do that.

It seems deosn't work in https://github.com/rolldown/rolldown/pull/2624/files#diff-67904986f81b9e6485293d52e55a19d1296ff8beef68cd7c4fee84fb26faea4a.


Maybe it's rolldown make some span invalid.

**Comment by IWANABETHATGUY** [comment]
> > > I tested it, here is a example and it works:
> > > ```
> > > /* #__PURE__ */ foo(a, b);
> > > ```
> > 
> > 
> > The `comments` option in codegen works well but may not work as your expectation, comments is not guaranteed to be generated in any position, both esbuild and swc can't do that.
> 
> It seems deosn't work in https://github.com/rolldown/rolldown/pull/2624/files#diff-67904986f81b9e6485293d52e55a19d1296ff8beef68cd7c4fee84fb26faea4a.
> 
> Maybe it's rolldown make some span invalid.

Neither esbuild or swc will generate that comments for you.
https://play.swc.rs/?version=1.8.0&code=H4sIAAAAAAAAA9PXUkjOz81NzSspNlTQ0ufSR%2FCNQPy0%2FHzrpMQiay4ufX1FBZ%2FU9MQchAaganQxdE0I44ASCkAJjUQdhSRNoBQAGxsxnHoAAAA%3D&config=H4sIAAAAAAAAA1VQu27DMAzc8xUG56AuPGTImiJbp36BoNCJCr1AUkCNwP9eSbYcZxPvwaPueeg6%2BGUN5%2B6Zn3mIihhpmzPCkxf1lxGQKSJrMlHg2FjhQgklrMi8ECCK7ijFhDx8DgM0nJTnMZDbJyTGLxyNx2ugi1XMV4P2xuveLYpQadn5MmRcDCQ%2FIZHGEvYQiXzue2T3wY8%2BLpbjy0DJi3FVq5IEp8RoWOn57Qc2BC66UVlebwBnvBmn%2Fek6uBzC%2FC4sUuXvdvMf2v66HVy4pUqupZdil65O8BK1sG0xGP5uztr4%2FA%2BBgdbpwAEAAA%3D%3D

https://hyrious.me/esbuild-repl/?version=0.23.0&b=e%00entry.js%00%2F*+comments1+*%2F%0A%2F*+comments2+*%2F%0Afoo%3Bbar%3B%0A%0A%2F%2F%21+Legal+comments1%0A%2F*%21+Legal+comments2+*%2F%0Afoo%3Bbar%3B%0A%0A%2F*+comments+*%2F+foo%28a%2C+b%29%3B&b=%00dep.js%00console.log%28%27dep%27%29%3B%0A%0Aexport+async+function+loadTs%28%29+%7B%0A++try+%7B%0A++++return+import%28%27.%2Flib.js%27%29.then%28%28m%29+%3D%3E+m.default%29%3B%0A++%7D+catch+%28e%29+%7B%0A++++console.error%28%27typescript+must+be+installed+to+use+%22native%22+functions%27%29%3B%0A++++throw+e%3B%0A++%7D%0A%7D%0Aexport+const+a+%3D+10%3B%0A&b=%00lib.js%00export+let+bar+%3D+123&o=%7B%0A++treeShaking%3A+true%2C%0A++bundle%3A+false%2C%0A++format%3A+%22esm%22%2C%0A++outdir%3A+%22dist%22%0A+%0A%7D

**Comment by hyf0** [comment]
Emmm. Looks like `/* #__PURE__ */ ` would be handled specially.

**Pull request merged**